src/token/aft34/extensions/payable_mint.rs

```diff
@@ -0,0 +1,156 @@
+// the following conditions:
+//
+// The above copyright notice and this permission notice shall be
+// included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+// EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+// NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+// LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+// WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+pub use crate::{
+    aft34,
+    traits::aft34::{extensions::metadata::*, extensions::payable_mint::*, *},
+};
+pub use aft34::{
+    AFT34Impl, BalancesManager as _, Internal as _, InternalImpl as _, Operator, Owner,
+};
+use ink::prelude::string::{String, ToString};
+use openbrush::{
+    contracts::ownable::*,
+    traits::{AccountId, Balance, Storage},
+};
+
+use super::metadata::AFT34MetadataImpl;
+
+#[derive(Default, Debug)]
+#[openbrush::storage_item]
+pub struct Data {
+    pub last_token_id: u64,
+    pub max_supply: u64,
+    pub price_per_mint: Balance,
+}
+
+#[openbrush::trait_definition]
+pub trait AFT34PayableMintImpl:
+    aft34::Internal
+    + aft34::aft34::AFT34Impl
+    + aft34::extensions::metadata::Internal
+    + aft34::extensions::payable_mint::Internal
+    + aft34::extensions::metadata::AFT34MetadataImpl
+    + aft34::InternalImpl
+    + Storage<Data>
+    + Storage<ownable::Data>
+{
+    fn mint(&mut self, to: AccountId, mint_amount: u64) -> Result<(), AFT34Error> {
+        self.check_value(Self::env().transferred_value(), mint_amount)?;
+        self.check_amount(mint_amount)?;
+
+        let next_to_mint = self.data::<Data>().last_token_id + 1; // first mint id is 1
+        let mint_offset = next_to_mint + mint_amount;
+
+        for mint_id in next_to_mint..mint_offset {
+            aft34::InternalImpl::_mint_to(self, to, Id::U64(mint_id))?;
+            self.data::<Data>().last_token_id += 1;
+        }
+
+        Ok(())
+    }
+
+    /// Withdraws funds to contract owner
+    #[ink(message)]
+    #[openbrush::modifiers(only_owner)]
+    fn withdraw(&mut self) -> Result<(), AFT34Error> {
+        let balance = Self::env().balance();
+        let current_balance = balance
+            .checked_sub(Self::env().minimum_balance())
+            .unwrap_or_default();
+        let owner = self.data::<ownable::Data>().owner.get().unwrap().unwrap();
+        Self::env()
+            .transfer(owner, current_balance)
+            .map_err(|_| AFT34Error::Custom(String::from("WithdrawalFailed")))?;
+        Ok(())
+    }
+
+    /// Set new value for the baseUri
+    #[ink(message)]
+    #[openbrush::modifiers(only_owner)]
+    fn set_base_uri(&mut self, uri: String) -> Result<(), AFT34Error> {
+        let id = AFT34Impl::collection_id(self);
```

Comment:
I dont get the use of collection_id() here as it always just return the account Id of the contract and NOT the id of the minted token.

Reply:
Isn't base_uri the same for the all contract? So we don't need to get token's ID here.

Reply:
The base_uri is supposed to be the part of the url of the storage that never change.

For exemple, a dapp is storing there assets on their server, so from the url https://mydapp.com/storage/nfts/1893, here the base_url is https://mydapp.com/storage/nfts as the only changing part is the ID.

Reply:
Then what is the point of having a `set_base_uri` method ? Shouldn't It just be in the constructor and that's it ?

Reply:
I think it safer if the owners are able to change this in case something happen on the actual base url storage or if it's moved on another domain name